File: packages/photo_manager_video_player/darwin/photo_manager_video_player/Sources/photo_manager_video_player/PhotoManagerVideoPlayerPlugin.swift
import Flutter
import UIKit

public class PhotoManagerVideoPlayerPlugin: NSObject, FlutterPlugin {
    static let channelName = "com.fluttercandies/photo_manager_video_player"

    private let registry: FlutterTextureRegistry
    private let messenger: FlutterBinaryMessenger
    private var players: [Int64: GalleryVideoPlayer] = [:]

    init(registry: FlutterTextureRegistry, messenger: FlutterBinaryMessenger) {
        self.registry = registry
        self.messenger = messenger
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = PhotoManagerVideoPlayerPlugin(
            registry: registrar.textures(),
            messenger: registrar.messenger()
        )
        let channel = FlutterMethodChannel(
            name: channelName, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    /// Releases every outstanding player when the engine tears down, so no
    /// `AVPlayer` or registered texture outlives the engine that owned it.
    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        releaseAll()
    }

    private func releaseAll() {
        players.values.forEach { $0.release() }
        players.removeAll()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]

        // A hot restart keeps the engine, so players from the previous isolate
        // would keep holding decoders with no Dart side left to dispose them.
        if call.method == "disposeAll" {
            releaseAll()
            return result(nil)
        }

        if call.method == "create" {
            guard let assetId = args["assetId"] as? String else {
                return result(FlutterError(
                    code: "playbackFailed", message: "Missing 'assetId'.", details: nil))
            }
            let player = GalleryVideoPlayer(
                registry: registry,
                messenger: messenger,
                localIdentifier: assetId,
                allowNetworkAccess: args["allowNetworkAccess"] as? Bool ?? true,
                looping: args["looping"] as? Bool ?? false,
                volume: args["volume"] as? Double ?? 1.0
            )
            players[player.textureId] = player
            return result(player.textureId)
        }

        guard let textureId = (args["textureId"] as? NSNumber)?.int64Value,
              let player = players[textureId] else {
            return result(FlutterError(
                code: "assetNotFound", message: "No player for the given texture.", details: nil))
        }

        switch call.method {
        case "play": player.play()
        case "pause": player.pause()
        case "seekTo": player.seek(toMilliseconds: (args["positionMs"] as? NSNumber)?.int64Value ?? 0)
        case "setVolume": player.setVolume(args["volume"] as? Double ?? 1.0)
        case "setLooping": player.setLooping(args["looping"] as? Bool ?? false)
        case "dispose": players.removeValue(forKey: textureId)?.release()
        default: return result(FlutterMethodNotImplemented)
        }
        result(nil)
    }
}

File: packages/photo_manager_video_player/android/src/main/kotlin/com/fluttercandies/photo_manager_video_player/PhotoManagerVideoPlayerPlugin.kt
package com.fluttercandies.photo_manager_video_player

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry

class PhotoManagerVideoPlayerPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var messenger: BinaryMessenger
    private lateinit var context: Context
    private lateinit var textures: TextureRegistry

    private val players = mutableMapOf<Long, GalleryVideoPlayer>()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        messenger = binding.binaryMessenger
        textures = binding.textureRegistry
        channel = MethodChannel(messenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        releaseAll()
        channel.setMethodCallHandler(null)
    }

    private fun releaseAll() {
        players.values.forEach { it.release() }
        players.clear()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        // A hot restart keeps the engine, so players from the previous isolate
        // would keep holding decoders with no Dart side left to dispose them.
        if (call.method == "disposeAll") {
            releaseAll()
            return result.success(null)
        }

        if (call.method == "create") {
            val assetId = call.argument<String>("assetId")
                ?: return result.error("playbackFailed", "Missing 'assetId'.", null)
            val player = GalleryVideoPlayer(
                context = context,
                messenger = messenger,
                surface = textures.createSurfaceProducer(),
                assetId = assetId,
                initialLooping = call.argument<Boolean>("looping") ?: false,
                initialVolume = call.argument<Double>("volume") ?: 1.0,
            )
            players[player.textureId] = player
            return result.success(player.textureId)
        }

        val textureId = call.argument<Number>("textureId")?.toLong()
        val player = players[textureId]
            ?: return result.error("assetNotFound", "No player for the given texture.", null)

        when (call.method) {
            "play" -> player.play()
            "pause" -> player.pause()
            "seekTo" -> player.seekTo(call.argument<Number>("positionMs")?.toLong() ?: 0L)
            "setVolume" -> player.setVolume(call.argument<Double>("volume") ?: 1.0)
            "setLooping" -> player.setLooping(call.argument<Boolean>("looping") ?: false)
            "dispose" -> players.remove(textureId)?.release()
            else -> return result.notImplemented()
        }
        result.success(null)
    }

    companion object {
        const val CHANNEL_NAME = "com.fluttercandies/photo_manager_video_player"
    }
}
